### PR TITLE
Fix/more data in train history

### PR DESCRIPTION
**Pull request by calbaker**
Added: 
```rust
    /// Current link containing head end (i.e. pulling locomotives) of train
    pub head_end_link_idx: u32,
    /// Offset from start of current link
    pub offset_in_link: si::Length,
```

to `TrainState` and corresponding code to store this in history

### CODE REVIEW

File: python/altrios/demos/speed_limit_train_sim_demo.py [hunk 1]
```diff
@@ -151,9 +149,62 @@
...
+    train_sim.history.speed_meters_per_second,
+)
+ax1[-1].set_xlabel('Time [hr]')
+ax1[-1].set_ylabel('Speed [m/s]')
+
+plt.suptitle("Speed Limit Train Sim Demo")
+plt.tight_layout()
+
+
+fig1, ax2 = plt.subplots(3, 1, sharex=True)
```
Comment (calbaker) [inline python/altrios/demos/speed_limit_train_sim_demo.py:177]: image

This demonstrates how to plot grade at front of train

File: python/altrios/demos/speed_limit_train_sim_demo.py [hunk 2]
```diff
@@ -151,9 +149,62 @@
...
+    train_sim.history.speed_meters_per_second,
+)
+ax1[-1].set_xlabel('Time [hr]')
+ax1[-1].set_ylabel('Speed [m/s]')
+
+plt.suptitle("Speed Limit Train Sim Demo")
+plt.tight_layout()
+
+
+fig2, ax2 = plt.subplots(3, 1, sharex=True)
```
Comment (calbaker) [inline python/altrios/demos/speed_limit_train_sim_demo.py:177]: image

File: python/altrios/demos/speed_limit_train_sim_demo.py [hunk 3]
```diff
@@ -151,9 +149,62 @@
 ax[-1].set_ylabel('Speed [m/s]')
 ax[-1].legend()
 plt.suptitle("Speed Limit Train Sim Demo")
+
+fig1, ax1 = plt.subplots(3, 1, sharex=True)
```
Comment (calbaker) [inline python/altrios/demos/speed_limit_train_sim_demo.py:153]: image

@SWRIganderson , I'd sure appreciate some serious scrutiny on this plot and similar plots in other contexts

**Comment by SWRIganderson** [comment]
@calbaker I have run this branch on my machine and inspected the results.  I believe it is working correctly.  I verified this by saving the data at 1 Hz and comparing the maximum link offset to the length of the link.  I did this for the first dozen links in the trip.  Everything checked out.

The thing that I also discovered is that the length calculation that is being done by QGIS may not be taking into account elevation or curvature. I have tracked this partially in the QGIS code, but I haven't run it to ground yet.  The link below is how far I have tracked it.  The screenshot is from QGIS.  They have two different ways of calculating distance.  The length in the network file is 7596.7 m.  We probably need to understand the differences between all of these calculations to make sure we are getting what we think we are.  It may be worth opening an issue maybe aroud this?

image


[https://github.com/qgis/QGIS/blob/master/src/analysis/processing/qgsalgorithmextractvertices.cpp](https://github.com/qgis/QGIS/blob/master/src/analysis/processing/qgsalgorithmextractvertices.cpp )

**Pull request merged**